app/receipt.go

```diff
@@ -0,0 +1,177 @@
+				if !found {
+					continue
+				}
+				spenderStr, found := GetAttributeValue(wasmEvent, "spender")
+				if !found {
+					continue
+				}
+				topics := []common.Hash{
+					ERC20ApprovalTopic,
+					app.GetEvmAddressAttribute(ctx, wasmEvent, "owner"),
+					app.GetEvmAddressAttribute(ctx, wasmEvent, "spender"),
+				}
+				res, err := app.WasmKeeper.QuerySmart(
+					ctx,
+					sdk.MustAccAddressFromBech32(contractAddr),
+					[]byte(fmt.Sprintf("{\"allowance\":{\"owner\":\"%s\",\"spender\":\"%s\"}}", ownerStr, spenderStr)),
+				)
+				if err != nil {
+					continue
+				}
+				allowanceResponse := &AllowanceResponse{}
+				if err := json.Unmarshal(res, allowanceResponse); err != nil {
+					continue
+				}
+				logs = append(logs, &ethtypes.Log{
+					Address: pointerAddr,
+					Index:   uint(len(logs)),
+					Topics:  topics,
+					Data:    common.BigToHash(allowanceResponse.Allowance.BigInt()).Bytes(),
+				})
+			}
+			continue
+		}
+	}
+	if len(logs) == 0 {
+		return
+	}
+	txHash := common.BytesToHash(checksum[:])
+	if response.EvmTxInfo != nil {
+		txHash = common.HexToHash(response.EvmTxInfo.TxHash)
+	}
+	var bloom ethtypes.Bloom
+	if r, err := app.EvmKeeper.GetTransientReceipt(ctx, txHash); err == nil && r != nil {
+		r.Logs = append(r.Logs, utils.Map(logs, evmkeeper.ConvertEthLog)...)
+		bloom = ethtypes.CreateBloom(ethtypes.Receipts{&ethtypes.Receipt{Logs: evmkeeper.GetLogsForTx(r)}})
+		r.LogsBloom = bloom[:]
+		_ = app.EvmKeeper.SetTransientReceipt(ctx, txHash, r)
+	} else {
+		bloom = ethtypes.CreateBloom(ethtypes.Receipts{&ethtypes.Receipt{Logs: logs}})
+		receipt := &evmtypes.Receipt{
+			TxType:           ShellEVMTxType,
+			TxHashHex:        txHash.Hex(),
+			GasUsed:          ctx.GasMeter().GasConsumed(),
+			BlockNumber:      uint64(ctx.BlockHeight()),
+			TransactionIndex: uint32(ctx.TxIndex()),
+			Logs:             utils.Map(logs, evmkeeper.ConvertEthLog),
+			LogsBloom:        bloom[:],
+			Status:           uint32(ethtypes.ReceiptStatusSuccessful), // we don't create shell receipt for failed Cosmos tx since there is no event anyway
+		}
+		sigTx, ok := tx.(authsigning.SigVerifiableTx)
+		if ok && len(sigTx.GetSigners()) > 0 {
+			// use the first signer as the `from`
+			receipt.From = app.EvmKeeper.GetEVMAddressOrDefault(ctx, sigTx.GetSigners()[0]).Hex()
+		}
+		_ = app.EvmKeeper.SetTransientReceipt(ctx, txHash, receipt)
+	}
+	if d := app.EvmKeeper.GetEVMTxDeferredInfo(ctx); d != nil {
+		app.EvmKeeper.AppendToEvmTxDeferredInfo(ctx, bloom, txHash, d.Surplus)
+	} else {
+		app.EvmKeeper.AppendToEvmTxDeferredInfo(ctx, bloom, txHash, sdk.ZeroInt())
+	}
+}
+
+func (app *App) GetEvmAddressAttribute(ctx sdk.Context, event abci.Event, attribute string) common.Hash {
+	addrStr, found := GetAttributeValue(event, attribute)
+	if found {
+		seiAddr, err := sdk.AccAddressFromBech32(addrStr)
+		if err == nil {
+			evmAddr := app.EvmKeeper.GetEVMAddressOrDefault(ctx, seiAddr)
```

Comment:
I believe it will, but just confirming:  this would work on a past block as well (ctx's height is honored here)?

Reply:
yeah all the states are accessed via ctx which will be versioned